Add support for FinMTEB benchmark

alt-glitch opened the pull request:
## Checklist

- [x] Run tests locally to make sure nothing is broken using `make test`. 
- [x] Run the formatter to format the code using `make lint`. 


### Adding datasets checklist
<!-- see also https://github.com/embeddings-benchmark/mteb/blob/main/docs/adding_a_dataset.md -->

**Reason for dataset addition**: 
- [FinMTEB Benchmark](https://github.com/yixuantt/FinMTEB)
- [Dataset](https://huggingface.co/FinanceMTEB):

Fixes #1267  

- [x] I have run the following models on the task (adding the results to the pr). These can be run using the `mteb -m {model_name} -t {task_name}` command.
  - [x] `sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2`
    - Ran only on `FiQAClassification` as of now.
  - [x] `intfloat/multilingual-e5-small`
    - Ran only on `FINAL` as of now.

- [ ] I have checked that the performance is neither trivial (both models gain close to perfect scores) nor random (both models gain close to random scores).
- [x] If the dataset is too big (e.g. >2048 examples), considering using `self.stratified_subsampling() under dataset_transform()`
- [x] I have filled out the metadata object in the dataset file (find documentation on it [here](https://github.com/embeddings-benchmark/mteb/blob/main/docs/adding_a_dataset.md#2-creating-the-metadata-object)).
- [x] Run tests locally to make sure nothing is broken using `make test`. 
- [x] Run the formatter to format the code using `make lint`.

alt-glitch commented:
Hey @Muennighoff @KennethEnevoldsen @isaac-chung!

Here's a WIP PR to close #1267.

I had a few questions/notes:

1. Should I run and get the results for all the tasks? 
2. Should the relevant PRs to `embeddings-benchmark/results` and `embeddings-benchmark/leaderboard` be made after merging this PR?
3. FiQA2018 is already in MTEB, so I have left that out from FinMTEB. Otherwise, there were no conflicting tasks.
4. The `Summarization` tasks are still pending. I have yet to look into the changes highlighted by @yixuantt in #1267 for summarization.

I'll add the summarization changes and make the PRs to results and leaderboard once this is done.
Is there anything else I'm missing out on?

isaac-chung commented:
Hi @alt-glitch , thanks for working on this!

1. Yes, I'd suggest running the whole thing on a small model mentioned in the paper like `all-MiniLM-L12-v2`, and only using the quickest settings as a sanity check, e.g. n_experiments=1 for classification.
2. Afterwards for the leaderboard yes. I'll leave the results repo part to @KennethEnevoldsen 
3. Sounds good.
4. I think it's ok to use the paper's URL or its [GitHub URL](https://github.com/yixuantt/FinMTEB/tree/main) as reference. Otherwise, there are individual references for each dataset mentioned in the paper. 
5. Re: summarization task, we can add column names as a class attributes to `AbstaskSummarization` like [the way we did in MIEB's AbsTaskImageClassification](https://github.com/embeddings-benchmark/mteb/blob/3a18fbdafa25696080fc4fa18c1875f64d6a4010/mteb/abstasks/Image/AbsTaskImageClassification.py#L35).

Let me know if anything is unclear.

KennethEnevoldsen commented:
Re. 2: PRs to `embeddings-benchmark/results` can be made after this PR. I don't believe a PR to `embeddings-benchmark/leaderboard` will be required once the new leaderboard is up and running as long as the benchmark is added to it is added to benchmarks.py

alt-glitch commented:
Thanks for the comments!

Some more info:
1. The `main_score` of PairClassification tasks needed to be fixed to `max_ap`.
2. Summarization tasks use `STSEvaluator`
3. Added reference_summaries_column and generated_summaries_column for specifying column names.
4. Added Summarization tasks.
5. Added some more Clustering tasks that I had missed out.
6. I'm currently taking a look at the results from the sanity run I did through the benchmark. All tasks ran fine.

I'm interested in helping with getting the results too!

cc @KennethEnevoldsen @isaac-chung

KennethEnevoldsen commented:
Not sure what is meant by:

> Summarization tasks use STSEvaluator

alt-glitch commented on mteb/tasks/Classification/zho/FinNSPClassification.py:
```diff
@@ -0,0 +1,25 @@
+        main_score="accuracy",
+    )
+
+    def dataset_transform(self):
+        if "sentence" in self.dataset:
```
Yep! This was left over while porting the class. Fixed.

alt-glitch commented on mteb/tasks/Classification/eng/SemEva2017Classification.py:
```diff
@@ -0,0 +1,24 @@
+        main_score="accuracy",
+    )
+
+    def dataset_transform(self):
+        self.dataset = self.dataset.rename_column("sentence", "text")
```
Added.

alt-glitch commented on mteb/tasks/Classification/eng/ESGClassification.py:
```diff
@@ -0,0 +1,21 @@
+    metadata = TaskMetadata(
+        name="ESGClassification",
+        description="A finance dataset performs sentence classification under the environmental, social, and corporate governance (ESG) framework.",
+        # reference="",
```
Added the references for tasks from the paper. For others, I added the FinMTEB reference itself.

alt-glitch commented on mteb/abstasks/AbsTaskSTS.py:
```diff
@@ -36,6 +36,8 @@ class AbsTaskSTS(AbsTask):
     """
 
     abstask_prompt = "Retrieve semantically similar text."
+    reference_summaries_column: str = "sentence1"
+    generated_summaries_column: str = "sentence2"
 
```
Renamed!

alt-glitch commented on mteb/tasks/Classification/eng/FOMCClassification.py:
```diff
@@ -0,0 +1,24 @@
+from mteb.abstasks.TaskMetadata import TaskMetadata
+
+
+class FOMCClassification(AbsTaskClassification):
```
Ah understood. Working on it.

alt-glitch commented on mteb/tasks/PairClassification/zho/AFQMCPairClassification.py:
```diff
@@ -0,0 +1,25 @@
+        main_score="max_ap",
+    )
+
+    def dataset_transform(self):
+        self.dataset = self.dataset.rename_column("sent1", "sentence1")
+        self.dataset = self.dataset.rename_column("sent2", "sentence2")
```
Done.

alt-glitch commented:
> Not sure what is meant by:
> 
> > Summarization tasks use STSEvaluator

Summarisation tasks here don't have human_summaries or relevance scores. So the Spearman correlation is calculated between summary and text. Hence the STSEvaluator is used.

See: https://github.com/yixuantt/FinMTEB/issues/2

---
1. Updated `AbsTaskSTS`, `AbsTaskSummarization`, `AbsTaskPairClassification` and the respective tasks to use configurable column names instead of `dataset_transform`
2. Added missing reference to the tasks.
3. I'm going to work on filling out the metadata for the tasks @KennethEnevoldsen. I'll update you once I'm done.

KennethEnevoldsen commented:
> Update: It's taking me a couple more days to fill out all the metadata fields for this benchmark as this seems to be mostly a manual process — reading the paper referenced for each dataset to understand and derive the date of dataset creation, annotation creators, and sample creation process since there are 64 datasets :)

Thanks for taking the time on this. I believe metadata is the only thing missing and then it can be reviewed and merged.

KennethEnevoldsen commented:
Moving this to v2.0.0 to avoid merge conflict in the future. I can resolve the current merge conflicts one metadata is added.

Muennighoff commented:
Would be great to get this in @alt-glitch in case you're still working on it! 🙌

alt-glitch commented:
Will push the metadata fields in soon!

Also any tips for how you guys go through so many papers and related datasets to add the metadata precisely?

isaac-chung commented:
@alt-glitch sorry for missing your question earlier. While personally I would do a few datasets metadata by hand, @imenelydiaker suggested using the hfapi client to automate some parts. I'll let her expand on that.

imenelydiaker commented:
Yes you can use HuggingFace APIs to get most of the metadata when they're available:

- For each `Dataset` object you can use `your_dataset_name.info` to get a `DatasetInfo` object containing the main metadata (mainly license, size and description): https://huggingface.co/docs/datasets/v2.1.0/en/package_reference/main_classes#datasets.DatasetInfo
- Use HF APIs to get information about splits (their size, name, and all the previous info + many other): https://huggingface.co/docs/dataset-viewer/en/info#get-dataset-information (I recommend having a script using this).

Clearly the info will not be available for all datasets, but it will save you some time. 🙂

isaac-chung reviewed (commented):
Nice work! Just a few questions.

Samoed commented on mteb/tasks/Classification/eng/FLSClassification.py:
```diff
@@ -7,15 +7,33 @@
+        task_subtypes=[],
+        license="not specified",
+        annotations_creators="derived",
+        descriptive_stats={
+            "num_samples": {"test": 1000},
+            "average_text_length": {"test": 187.923},
+            "unique_labels": {"test": 3},
+            "labels": {
+                "test": {
+                    "2": {"count": 292},
+                    "1": {"count": 539},
+                    "0": {"count": 169},
+                }
+            },
+        },
```
Now, the descriptive statistics are saved as JSON and do not need to be added in this way.

alt-glitch replied:
ah gotcha. let me rework this.

thanks for the heads-up!

Samoed commented on mteb/tasks/Summarization/zho/FinEvasum.py:
```diff
@@ -19,8 +19,8 @@ class FinEvasum(AbsTaskSTS):
         main_score="cosine_spearman",
     )
-    reference_summaries_column = "text"
-    generated_summaries_column = "summary"
```
I think these tasks should be moved out `Summarization` folder to `STS`

Samoed commented:
They've submitted paper https://huggingface.co/papers/2502.10990

KennethEnevoldsen commented:
And the code:
https://github.com/yixuantt/FinMTEB

seems like most tasks can simply be ported over

KennethEnevoldsen commented:
@alt-glitch, it seems like this PR has gotten stale. Do we still want to finalize this merge?

KennethEnevoldsen commented:
Will close this for now as it seems to have gotten stale
